x-pack/legacy/plugins/code/server/distributed/multinode/non_code_node_adapter.ts

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Wreck from '@hapi/wreck';
+import util from 'util';
+import Boom from 'boom';
+import { Request } from 'hapi';
+import * as http from 'http';
+import { DEFAULT_SERVICE_OPTION, HandlerAdapter, ServiceRegisterOptions } from '../handler_adpter';
+import { ResourceLocator } from '../resource_locator';
+import { ServiceDefinition, ServiceHandlerFor, ServiceMethodMap } from '../service_definition';
+import { CodeNodeResourceLocator } from './code_node_resource_locator';
+import { CodeNodeEndpoint } from './code_node_endpoint';
+import { RequestPayload } from './code_node_adapter';
+import { Logger } from '../../log';
+
+const pickHeaders = ['authorization'];
+
+function filterHeaders(originRequest: Request) {
+  const result: { [name: string]: string } = {};
+  for (const header of pickHeaders) {
+    if (originRequest.headers[header]) {
+      result[header] = originRequest.headers[header];
+    }
+  }
+  return result;
+}
+
+export class NonCodeNodeAdapter implements HandlerAdapter {
+  handlers: Map<any, any> = new Map<any, any>();
+
+  constructor(private readonly codeNodeUrl: string, private readonly log: Logger) {}
+
+  locator: ResourceLocator = new CodeNodeResourceLocator(this.codeNodeUrl);
+
+  getService<def extends ServiceDefinition>(serviceDefinition: def): ServiceMethodMap<def> {
+    const serviceHandler = this.handlers.get(serviceDefinition);
+    if (!serviceHandler) {
+      // we don't need implementation code for service, so we can register here.
+      this.registerHandler(serviceDefinition, null);
+    }
+    return serviceHandler as ServiceMethodMap<def>;
+  }
+
+  registerHandler<def extends ServiceDefinition>(
+    serviceDefinition: def,
+    serviceHandler: ServiceHandlerFor<def> | null,
```

Comment:
nit: a brief comment on that serverHandler will always be `null` here since it will be overridden by the request forwarding.